### PR TITLE
Ckpt01

### PR DESCRIPTION
**Pull request by dawsonc623**

### CODE REVIEW

File: 01-data-structures/01-introduction-to-data-structures/screen/pixel.rb [hunk 1]
```diff
@@ -11,11 +11,17 @@ class Pixel
...
+    self.green = validate_color(green)
+    self.blue = validate_color(blue)
+
+    self.x, self.y = x, y
   end
 
   private
 
   def validate_color(color)
+    [0, [color, 255].min].max
```
Comment (dawsonc623) [inline 01-data-structures/01-introduction-to-data-structures/screen/pixel.rb]: Clever way of handling it. I wonder what performance implications there are to the array constructions.

File: 01-data-structures/01-introduction-to-data-structures/screen/screen.rb [hunk 2]
```diff
@@ -6,18 +6,24 @@ class Screen
...
   end
 
   def at(x, y)
+    inbounds(x, y) ? matrix[x][y] : nil
   end
 
   private
 
   def inbounds(x, y)
+    ((0..width).include?(x)) && ((0..height).include?(y))
```
Comment (dawsonc623) [inline 01-data-structures/01-introduction-to-data-structures/screen/screen.rb]: Clever once again. I am not solid on the implementation details of ranges in Ruby, but depending on how `include?` works, there could be some performance concerns. However, I can think of a number of viable implementations that would not be too bad performance-wise.

File: 01-data-structures/01-introduction-to-data-structures/line/line_answers.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+Used an Array, a linear, ordered arrangement of elements and suggested by initialize method of the exercise code
+
+1) a) joining the line means to simply append the new member to the array
+   b) searching for a particular member requires iterating through each element until a match is found
+   c) removing a particular member, requires first searching (see above) and then setting element to be empty or nil (though probably should just remove the element to shorten the line
```
Comment (dawsonc623) [inline 01-data-structures/01-introduction-to-data-structures/line/line_answers.txt]: I would agree with just removing the element, but the assignment did not specify how it should be handled.

File: 01-data-structures/01-introduction-to-data-structures/screen/screen_answers.txt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+Used an Array of Arrays to act as a two-dimensional array (or matrix)
+
+1) a) inserting a pixel can be done directly to the Array of Arrays by simply referring to their position in the arrays using x and y
+   b) searching for a pixel by x and y position is done by simply accessing the same element indices within the two-dimensional array
+   c) removing a particular pixel can be done by reseting the element at its position with nil
```
Comment (dawsonc623) [inline 01-data-structures/01-introduction-to-data-structures/screen/screen_answers.txt]: A remove method of some variety may do well here.

**Comment by rdnewman** [comment]
Thanks for resetting the PR to my repo.  Sorry I didn't get back to it sooner.   I appreciate the comments.   So you got me curious.  

It turns out a simple ternary treatment (e.g., `color < 0 ? 0 : (color > 255 ? 255 : color)`) for how I did `[0, [color, 255].min].max` would be much faster (about 7X faster).  

The range approach of `((0..width).include?(x)) && ((0..height).include?(y))` was faster than a few other ways I tried, but the fairly direct approach of `(x < 0 ? 0 : (x > width ? width : x)) && (y < 0 ? 0 : (y > height ? height : y))` was indeed faster, by about 1/3, so not as dramatic.

**Comment by rdnewman** [comment]
I added my benchmarking in case you were curious.

**Pull request merged**